wainlog/adapters/api/views.py

```diff
@@ -60,50 +75,88 @@ def root() -> str | Response:
+    if not current_user.is_anonymous:
         return redirect(url_for("wainlog.root"))
 
-    form = forms.Login()
+    session["oauth2_state"] = secrets.token_urlsafe(16)
+
+    query_string = urlencode(
+        {
+            "client_id": current_app.config["GOOGLE_OAUTH_CLIENT_ID"],
+            "redirect_uri": url_for(
+                "wainlog.oauth2_callback", provider=provider, _external=True
+            ),
+            "response_type": "code",
+            "scope": " ".join(current_app.config["GOOGLE_OAUTH_SCOPES"]),
+            "state": session["oauth2_state"],
+        }
+    )
 
-    if form.validate_on_submit():
-        user = persister.get_user_db_from_username(
-            session=g.session, username=form.username.data
-        )
+    return redirect(
+        current_app.config["GOOGLE_OAUTH_AUTHORIZE_URL"] + "?" + query_string
+    )
 
-        if user is None or user.check_password(form.password.data) is False:
-            flash("Invalid username or password")
-            return redirect(url_for("wainlog.login"))
-        login_user(user)
 
+@bp.route("/callback/<provider>", methods=["GET"])
+def oauth2_callback(provider: str) -> Response:  # noqa: C901
+    if not current_user.is_anonymous:
         return redirect(url_for("wainlog.root"))
 
-    return render_template("login.html", form=form)
+    if "error" in request.args:
+        for k, v in request.args.items():
+            if k.startswith("error"):
+                flash(f"{k}: {v}")
+        return redirect(url_for("index"))
+
+    if request.args["state"] != session.get("oauth2_state"):
+        abort(401)
+
+    if "code" not in request.args:
+        abort(401)
+
+    response = requests.post(
+        current_app.config["GOOGLE_OAUTH_TOKEN_URL"],
+        data={
+            "client_id": current_app.config["GOOGLE_OAUTH_CLIENT_ID"],
+            "client_secret": current_app.config["GOOGLE_OAUTH_CLIENT_SECRET"],
+            "code": request.args["code"],
+            "grant_type": "authorization_code",
+            "redirect_uri": url_for(
+                "wainlog.oauth2_callback", provider=provider, _external=True
+            ),
+        },
+        headers={"Accept": "application/json"},
+    )
+    if response.status_code != 200:
+        abort(401)
+    oauth2_token = response.json().get("access_token")
+    if not oauth2_token:
+        abort(401)
+
+    response = requests.get(
+        current_app.config["GOOGLE_OAUTH_USERINFO_URL"],
+        headers={
+            "Authorization": "Bearer " + oauth2_token,
+            "Accept": "application/json",
+        },
+    )
+    if response.status_code != 200:
+        abort(401)
+    email = response.json()["email"]
+
+    user = persister.get_user_db_from_email(g.session, email)
+    if user is None:
+        user = model.User(email=email, username=email.split("@")[0])
+        persister.add_user(g.session, user)
+        g.session.commit()
+
+    login_user(user)
```

Comment:
Now you're logging them in, is there anywhere you need to decorate or check that the user is authenticated?

I see on line 32 you check in this file, you might want to flip the conditional and keep the happy path down the margin. In general you might want to make an `@auth_required` decorator if one doesn't already exist for any pages that should only be visible when logged in.

As for session timeouts, the timeout here is controlled by flask-login's session timeout, because that's what's actually used to auth the user.

Reply:
> Now you're logging them in, is there anywhere you need to decorate or check that the user is authenticated?

Under a multi-page setup normally, then yea you'd decorate views with `@flask_login.login_required`. But because mine is single-page, I'm not sure I really have a choice (I could decorate with `@login_required` and have conditional logic to bypass that maybe, but I suspect that would be quite clunky). What do you reckon?

> As for session timeouts, the timeout here is controlled by flask-login's session timeout, because that's what's actually used to auth the user

Niceee, thanks 🎉

Reply:
I guess I could create a duplicate view which requires login and renders the same template, and the split the logic up between the two of them. Maybe that's neater?

Reply:
You wouldn't be able to bind them both to the same route.

I'd probably just flip the conditional on line 32, and bin off the un-authed branch early, keeping the main logic down the margin.

Reply:
> You wouldn't be able to bind them both to the same route

Yaaa sorry my wording was bad here - I basically just meant two views which render the same template - one for unauthed, and one for authed.

> I'd probably just flip the conditional on line 32, and bin off the un-authed branch early, keeping the main logic down the margin

Ahhh ye that's neater, shall do that thanks!

Reply:
https://github.com/figchutney/wainlog/pull/1/commits/bc2a5307902a678b9ea7763e8eaf5629d3e1aa83